# Feature extraction DEMO

*"Can we extract age & ethnicity from PFD reports?"*

**Context:** There are no age or ethnicity fields in PFD reports. However, anecdotally, many coroners seem to note age in the "Investigation & Inquest" or "Circumstances of Death" sections of the report.

This demo considers the extent to which age & ethnicity can be extracted from these reports using PFD Toolkit's `Extractor` class.

In [4]:
from pfd_toolkit import load_reports, LLM, Extractor
from dotenv import load_dotenv
import os
from pydantic import BaseModel

# Load reports sample
reports = load_reports(n_reports=150)

# Load OpenAI API key
load_dotenv("api.env")
openai_api_key = os.getenv("OPENAI_API_KEY")

# Initialise LLM client
llm_client = LLM(api_key=openai_api_key, max_workers=30)

# Initialise Pydantic model
class DemoFeatures(BaseModel):
    age: int
    ethnicity: str

# Initialise extraction instructions
feature_instructions = {
    "age": "Age of the deceased",
    "ethnicity": "Ethnicity of the deceased",
}

extractor = Extractor(
    llm=llm_client,
    feature_model=DemoFeatures,
    feature_instructions=feature_instructions,
    reports=reports,
    include_investigation=True,
    include_circumstances=True,
)

result_df = extractor.extract_features()
result_df


LLM: Extracting features: 100%|██████████| 150/150 [00:05<00:00, 27.32it/s]


URL         ID       Date  \
0    https://www.judiciary.uk/prevention-of-future-...  2025-0248 2025-05-28   
1    https://www.judiciary.uk/prevention-of-future-...  2025-0243 2025-05-27   
2    https://www.judiciary.uk/prevention-of-future-...  2025-0244 2025-05-27   
3    https://www.judiciary.uk/prevention-of-future-...  2025-0245 2025-05-27   
4    https://www.judiciary.uk/prevention-of-future-...  2025-0246 2025-05-27   
..                                                 ...        ...        ...   
145  https://www.judiciary.uk/prevention-of-future-...  2025-0096 2025-02-19   
146  https://www.judiciary.uk/prevention-of-future-...  2025-0092 2025-02-18   
147  https://www.judiciary.uk/prevention-of-future-...  2025-0099 2025-02-18   
148  https://www.judiciary.uk/prevention-of-future-...  2025-0098 2025-02-18   
149  https://www.judiciary.uk/prevention-of-future-...  2025-0091 2025-02-17   

           CoronerName                     Area  \
0         Clare Bailey  Teesside and Hartlepool   
1       Andrew Cousins        Blackpool & Fylde   
2       Peter Merchant      West Yorkshire West   
3        Nadia Persaud              East London   
4       Rebecca Sutton    Durham and Darlington   
..                 ...                      ...   
145        Susan Ridge                   Surrey   
146  Caroline Saunders                    Gwent   
147       Sarah Bourke       Inner North London   
148       Sarah Bourke       Inner North London   
149      Paul Appleton  Teesside and Hartlepool   

                                              Receiver  \
0    1 Department of Health and Social Care 2 Chief...   
1                      BARCHESTER HEALTHCARE LIMITED 1   
2          1 , Chief Constable West Yorkshire Police 1   
3    , Chief Executive Officer, Barts Health NHS Fo...   
4    1. Deputy Chief Constable , Durham Constabular...   
..                                                 ...   
145  Chief Executive Surrey and Sussex Healthcare N...   
146                                   Welsh Parliament   
147  1. , Secretary of State for Justice, Ministry ...   
148  1. , Secretary of State for Justice, Ministry ...   
149  1 DEPARTMENT OF HEALTH. 2 NHS ENGLAND. 3 NHS N...   

                               InvestigationAndInquest  \
0    Mr Dean Bradley died on 15 th October 2021 at ...   
1    On 30 April 2025 and 23 May 2025, at an inques...   
2    On 15 February 2024 the death of Paul Andrew A...   
3    On the 13 June 2024 I commenced an investigati...   
4    On 7 January 2025 an investigation into the de...   
..                                                 ...   
145  An inquest into Mrs Rodgers death was opened o...   
146  On 14/3/2024, an investigation was opened touc...   
147  On 4 October 2023, Assistant Coroner Smith com...   
148  On 19 October 2022, I commenced an investigati...   
149  On 18 September 2024 I commenced an investigat...   

                                  CircumstancesOfDeath  \
0    At approximately 0300 on 15 th October 2021 Mr...   
1    I returned the following in box 4 of the Recor...   
2    As identified above, Paul Alexander had a long...   
3    Abdirahman Afrah began to suffer from chest pa...   
4    The Deceased had a long history of mental heal...   
..                                                 ...   
145  A narrative conclusion was recorded at Box 4 o...   
146  On 19/2/2024, Jeffrey Martin Tyler called the ...   
147  Mr Bainborough lived in supported living accom...   
148  Mrs Mohamed developed mental health problems a...   
149  Following a call to the 111 Service at 22:14 o...   

                                      MattersOfConcern             age  \
0    During the course of the investigation my inqu...  N/A: Not found   
1    During the course of the inquest the evidence ...  N/A: Not found   
2    During the course of the investigation my inqu...  N/A: Not found   
3    During the course of the inquest the evidence ...              17   
4    During the 

In [2]:
result_df.to_csv('../data/age_extracted.csv')

We can see that age seems generally well extracted. However, ethnicity does not appear to be well extracted. Let's unpack this a bit more...

In [16]:
# For all ages print True, for all missing data print False
print(result_df["age"].ne("N/A: Not found").value_counts())

# Print all unique values counts for ethnicity
print("\n\n")
print(result_df["ethnicity"].value_counts())

age
True     110
False     40
Name: count, dtype: int64



ethnicity
N/A: Not found    149
Black               1
Name: count, dtype: int64


Conclusions: 

1. In our sample, it was possible to extract age as a feature in 110 reports, with 40 reports possibly not recording this information.

2. Our sample showed that ethnicity was recorded just once, with 149 reports possibly not recording this information. In this one instance, the model correctly extracted this information.